common/crypto_utils/avalon_crypto_utils/signature.py

```diff
@@ -323,35 +323,69 @@ def generate_client_signature(
         return input_json_str, SignatureStatus.PASSED
 
 # -----------------------------------------------------------------------------
-    def verify_signature(self, input_json, verification_key):
+    def verify_signature(self, wo_response, verification_key,
+                         requester_nonce):
         """
         Function to verify the signature received from the enclave
         Parameters:
-            - input_json is dictionary contains payload returned by the
+            @param wo_response - dictionary contains payload returned by the
               Worker Service in response to successful workorder submit request
               as per Trusted Compute EEA API 6.1.2 Work Order Result Payload
-            - verification_key is ECDSA/SECP256K1 public key used to verify
+            @param verification_key - ECDSA/SECP256K1 public key used to verify
               signatures created by the Enclave.
+            @param requester_nonce - requester generated nonce passed in work
+            order request. Required in 2 step verification.
         Returns enum type SignatureStatus
         """
-
-        input_json_params = input_json
-
-        nonce = (input_json_params['workerNonce']).encode('UTF-8')
-        signature = input_json_params['workerSignature']
-
+        v_key = None
+        # if verification_key present in work order response
+        # then do 2 step verification
+        # step1 - The verification key signature from the
+        # response is verified using worker’s public verification key
+        # (aka KME's verification key)
+        if "extVerificationKey" in wo_response:
+            decrypted_v_key = crypto.base64_to_byte_array(
+                wo_response["extVerificationKey"])
+            v_key = crypto.byte_array_to_string(decrypted_v_key)
+            if requester_nonce is None:
+                logger.info("Missing requester_nonce argument")
+                SignatureStatus.FAILED
+            concat_string = v_key + requester_nonce
+            v_key_sig = wo_response["extVerificationKeySignature"]
+            v_key_hash = crypto.compute_message_hash(
+                bytes(concat_string, 'UTF-8'))
+            try:
+                _verifying_key = crypto.SIG_PublicKey(verification_key)
+            except Exception as error:
+                logger.info("Error in work order "
+                            "response verification key : %s", error)
+                return SignatureStatus.INVALID_VERIFICATION_KEY
+            decoded_v_key_sig = crypto.base64_to_byte_array(v_key_sig)
+            sig_result = _verifying_key.VerifySignature(
+                v_key_hash,
+                decoded_v_key_sig)
+            if sig_result != 1:
+                logger.info("Error in verifying "
+                            "verification key signature")
+                return SignatureStatus.FAILED
+        else:
```

Comment:
Can we move else part of another function. This will help reduce complexity of this function. Can we make requester_nonce optional. That way existing applications need not change. requestor_nonce should be present only for extVerificationKeySignature. something like below

```
def verify_signature(self, wo_response, verification_key,requester_nonce=None):

  if "extVerificationKey" in wo_response:
      self.verify_worker_signature_ext(wo_response, verification_key,requester_nonce)
  else:
      self.verify_worker_signature(wo_response,verification_key)
```

Reply:
What if worker pool missed that argument? One function handling both the cases(singleton and worker pool) One API

Reply:
If worker pool doesn't pass `requestor_nonce` argument then return error in `verify_worker_signature_ext()`. For singleton case requestor_nonce is not required. We can still keep one function to handle both cases